datalad/local/subdatasets.py

```diff
@@ -0,0 +1,438 @@
+            By default subdatasets are reported regardless of their
+            status""",
+            constraints=EnsureBool() | EnsureNone()),
+        recursive=recursion_flag,
+        recursion_limit=recursion_limit,
+        contains=Parameter(
+            args=('--contains',),
+            metavar='PATH',
+            action='append',
+            doc="""limit report to the subdatasets containing the
+            given path. If a root path of a subdataset is given the last
+            reported dataset will be the subdataset itself.[CMD:  This
+            option can be given multiple times CMD][PY:  Can be a list with
+            multiple paths PY], in which case datasets will be reported that
+            contain any of the given paths.""",
+            constraints=EnsureStr() | EnsureNone()),
+        bottomup=Parameter(
+            args=("--bottomup",),
+            action="store_true",
+            doc="""whether to report subdatasets in bottom-up order along
+            each branch in the dataset tree, and not top-down."""),
+        set_property=Parameter(
+            args=('--set-property',),
+            metavar=('NAME', 'VALUE'),
+            nargs=2,
+            action='append',
+            doc="""Name and value of one or more subdataset properties to
+            be set in the parent dataset's .gitmodules file. The property name
+            is case-insensitive, must start with a letter, and consist only
+            of alphanumeric characters. The value can be
+            a Python format() template string wrapped in '<>' (e.g.
+            '<{gitmodule_name}>').
+            Supported keywords are any item reported in the result properties
+            of this command, plus 'refds_relpath' and 'refds_relname':
+            the relative path of a subdataset with respect to the base dataset
+            of the command call, and, in the latter case, the same string with
+            all directory separators replaced by dashes.[CMD:  This
+            option can be given multiple times. CMD]""",
+            constraints=EnsureStr() | EnsureNone()),
+        delete_property=Parameter(
+            args=('--delete-property',),
+            metavar='NAME',
+            action='append',
+            doc="""Name of one or more subdataset properties to be removed
+            from the parent dataset's .gitmodules file.[CMD:  This
+            option can be given multiple times. CMD]""",
+            constraints=EnsureStr() | EnsureNone()))
+
+    @staticmethod
+    @datasetmethod(name='subdatasets')
+    @eval_results
+    def __call__(
+            path=None,
+            dataset=None,
+            fulfilled=None,
+            recursive=False,
+            recursion_limit=None,
+            contains=None,
+            bottomup=False,
+            set_property=None,
+            delete_property=None):
+        # no constraints given -> query subdatasets under curdir
+        if path is None and dataset is None:
```

Comment:
From the command line, `path` won't ever come in as None.

This means that the test

```python
with chpwd(text_type(ds.pathobj / 'subdir')):
    # imitate cmdline invocation w/ no dataset argument
    # -> curdir limits the query, when no info is given
    eq_(subdatasets(dataset=None,
                    path=None,
                    recursive=True,
                    result_xfm='paths'),
        [text_type(ds.pathobj / dots)]
    )
```

doesn't actually match the command-line behavior:

```
% datalad subdatasets -r
subdataset(ok): sub dataset1 (dataset)
subdataset(ok): sub dataset1/2 (dataset)
subdataset(ok): sub dataset1/sub sub dataset1 (dataset)
subdataset(ok): sub dataset1/sub sub dataset1/2 (dataset)
subdataset(ok): sub dataset1/sub sub dataset1/subm 1 (dataset)
subdataset(ok): sub dataset1/subm 1 (dataset)
subdataset(ok): subdir/.lots.of.dots. (dataset)
action summary:
  subdataset (ok: 7)
```

Perhaps this should be `if not path ...` instead.

Reply:
Indeed. Fixed.